_source/_posts/2022-02-25-thymeleaf-security.md

```diff
@@ -0,0 +1,670 @@
+                .toString().contains("An expected CSRF token cannot be found");
+    }
+
+    @Test
+    void testPostQuiz() throws Exception {
+        this.client.mutateWith(csrf()).mutateWith(mockOidcLogin())
+                .post().uri("/quiz")
+                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
+                .exchange().expectStatus().isOk();
+    }
+
+    @Test
+    void testGetQuiz_noAuth() throws Exception {
+        this.client.get().uri("/quiz").exchange().expectStatus().is3xxRedirection();
+    }
+
+    @Test
+    void testGetQuiz() throws Exception {
+        this.client.mutateWith(mockOidcLogin())
+                .get().uri("/quiz").exchange().expectStatus().isOk();
+    }
+}
+```
+
+In the test class above, the first test `testPostQuiz_noCSRFToken()` verifies the quiz cannot be submitted without the CSRF token, even if the user did sign in. In the second test `testPostQuiz()`, the CSRF token is added to the mock request with `mutateWith(csrf())`, so the expected response status is HTTP 200 OK. The third test `testGetQuiz_noAuth` verifies a request of the quiz will be redirected (to Okta sign-in form) if the user is not authenticated. And the last test `testGetQuiz()` verifies the quiz can be accessed if the user has been authenticated with an OIDC login.
+
+Before running the application, as `quiz` is not a standard scope or a scope defined in Okta, you have to define it for the default authorization server. Sign in to the Okta dashboard, and in the left menu, go to **Security > API**, the choose the **default** authorization server. In the **Scopes** tab, click **Add Scope**. Set the scope name as `quiz` and add a description, leave all the remaining fields with default values and click on **Create**. Now the `quiz` scope can be required during the OIDC login.
+
+{% img blog/thymeleaf-security/add-scope.png alt:"Add Scope Page" width:"800" %}{: .center-image }
+
+
+Run the application without adding the `quiz` scope to the `application.yml` file, sign in and you should not see the quiz link yet. If you make a GET request with the browser to http://localhost:8080/quiz, the response will be 403 Forbidden.
+
+Now add `quiz` to the list of scopes in the Okta configuration in `application.yml`. The final configuration should look like this:
+
+```yml
+spring:
+  security:
+    oauth2:
+      client:
+        provider:
+          okta:
+            user-name-attribute: email
+
+okta:
+  oauth2:
+    issuer: https://{yourOktaDomain}/oauth2/default
+    client-id: {clientId}
+    client-secret: {clientSecret}
+    scopes:
+      - email
+      - openid
+      - profile
+      - quiz
+```
+
+Run the application again and you should see the quiz link "Visit the **Thymeleaf Quiz** to test Cross-Site Request Forgery (CSRF) protection". Click the link and the quiz will be displayed:
+
+
+{% img blog/thymeleaf-security/quiz-form.png alt:"Thymeleaf Quiz Form" width:"800" %}{: .center-image }
```

Comment:
The font seems a little small in the screenshot. It looks great in the developer console! Maybe bump the fonts a bit in the browser and retake the screenshot?

Reply:
Retook the screenshot